[core] reuse `attachComment` logic and export from core API

pieterv opened the pull request:
## Description

Move all comment attachment setup logic to a central `attachComments` function. We can then reuse this within `multiparsers.js`. Additionally export the `attachComments` API under the `prettier.utils` object.

General notes:
- `attachComments` now returns the cached comments from `options` along with the newly attached comments, this enables that `ensureAllCommentsPrinted` to correctly check attached comments are printed even if attachment is manually triggered outside of the prettier call.
- `attachComments` is merged with the `utils` to avoid a circular dependency between `comments` and `utils`.

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
   - No will do this as a followup or after agreeing on the direction.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

thorn0 commented:
Did you see #8122? Eventually we'd like to stop mutating the AST. So I'm not comfortable with exposing this API.

thorn0 commented:
On second thought, this change doesn't really expose how exactly comments are associated with nodes. If the `comments` property gets replaced with weak maps, it won't break. So it seems like my concern is irrelevant.

pieterv commented:
> Did you see #8122?

No i hadn't, that was interesting, thanks for the link. I would love to see a `prettier.formatAST(ast)` style API, the current `parser` style API is pretty clunky as you cannot cancel the printing after parsing (which is an important use case for codemods).

> If the comments property gets replaced with weak maps, it won't break.

Yes this makes sense to me. I guess the only caveat is when mutating the AST tools should use the prettier helpers for mutating comments rather than mutating the `comments` array (likely the same change you will need to make internally to prettier). However we might need remove + read versions of `util.addTrailingComment/addLeadingComment/addDanglingComment` to make this possible e.g. moving comments from one node to another is a comment operation.

pieterv commented:
Added `attachComments` API to `standalone.js` to fix the failing tests.

pieterv commented:
Derp, actually look into and the fix failing test. Verified via `yarn test tests/integration/__tests__/util-shared.js`, i just removed the check i added since `attachComments` is added to utils in `index.js` and verified in the other test i added.

pieterv commented:
Just following up here. Is there any changes needed to this? Happy to keep iterating if needed.
